<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

# Project 1: World Development Statistics

--- 
# Part 1

Part 1 requires knowledge of basic Python.

---

## Problem Statement

Decide on your problem statement that will guide your analysis for this project. For guidelines, sample prompts, or inspiration, check out the README.

**To-Do:** *Replace this cell with your problem statement.*

Doctors-Go-Anywhere have reached out after their recent funding, as they are looking to expand their operations. They are wanting to indentify the Top 5 regions globally that have the least Healthcare access and would benefit from their services.

### Contents:
- [Background](#Background)
- [Data Import & Cleaning](#Data-Import-and-Cleaning)
- [Exploratory Data Analysis](#Exploratory-Data-Analysis)
- [Data Visualization](#Visualize-the-Data)
- [Conclusions and Recommendations](#Conclusions-and-Recommendations)

## Background

For our first project, we're going to take a look at World Development Statistics from [Gapminder](https://www.gapminder.org/about/), an independent Swedish foundation that aims to make data about the world more accessible and reliable. A good introduction on Gapminder is this [Ted Talk](https://www.ted.com/talks/hans_rosling_the_best_stats_you_ve_ever_seen) from Hans Rosling, which also shows how effective data visualization can be for your audience.

**To-Do:** *Fill out this cell (or edit the above cell) with any other background or information that is necessary for your problem statement.*

Additional will also be pulled from world health organization. We will also be looking into see the number of existing doctors that there are in regions and see how much governments are investing into their healthcare to see how that collerates into in their position on the rankings.


### Choose your Data

There are 3 datasets included in the [`data`](./data/) folder for this project. You are required to pick **at least two** of these to complete your analysis. Feel free to use more than two if you would like, or add other relevant datasets you find online. 

**HINT: There are more datasets available on the Gapminder website**

* [`population.csv`](./data/population.csv): Population by Country
* [`life_expectancy.csv`](./data/life_expectancy.csv): Life Expectancy by Country
* [`gni_per_cap_atlas_method_con2021.csv`](./data/gni_per_cap_atlas_method_con2021.csv): Gross National Income (GNI) per capita in current US dollars

**To-Do:** *Fill out this cell with the datasets you will use for your analysis. Write a brief description of the contents for each dataset that you choose.*

Alongside the population & life expectancy data files, I will be pulling data on the number of doctors, health aid given and total health spending.


### Outside Research

Based on your problem statement and your chosen datasets, spend some time doing outside research on state policies or additional information that might be relevant. Summarize your findings below. If you bring in any outside tables or charts, make sure you are explicit about having borrowed them. If you quote any text, make sure that it renders as being quoted. **Make sure that you cite your sources.**

**To-Do:** *Fill out this cell with outside research or any additional background information that will support your analysis.*

### Coding Challenges

1. Manually calculate mean:

    Write a function that takes in values and returns the mean of the values. Create a list of numbers that you test on your function to check to make sure your function works!
    
    *Note*: Do not use any mean methods built-in to any Python libraries to do this! This should be done without importing any additional libraries.

In [6]:
# Code:
list_numbers = [56, 577, 343, 332, 566, 21, 96]
mean = sum(list_numbers)/len(list_numbers); 

2. Manually calculate standard deviation:

    The formula for standard deviation is below:

    $$\sigma = \sqrt{\frac{1}{n}\sum_{i=1}^n(x_i - \mu)^2}$$

    Where $x_i$ represents each value in the dataset, $\mu$ represents the mean of all values in the dataset and $n$ represents the number of values in the dataset.

    Write a function that takes in values and returns the standard deviation of the values using the formula above. Hint: use the function you wrote above to calculate the mean! Use the list of numbers you created above to test on your function.
    
    *Note*: Do not use any standard deviation methods built-in to any Python libraries to do this! This should be done without importing any additional libraries.

In [7]:
# Code:
((sum((num - mean)**2 for num in list_numbers))/len(list_numbers)) ** 0.5


216.35081111331144

3. Data cleaning function:
    
    Write a function that takes in a string consisting a number and an abbreviated place value (ex. '5M', '45K'), and converts that string to an integer. For example, inputting '5M' in your function should return 5000000, '45K' should return 45000, etc. Make sure to test your function to make sure it works!

You will use these functions later on in the project!

In [28]:
# Code:
def string_tonum(org_fig):
    keys = {'K': 1000, 'M': 1000000, 'B': 1000000000, '0': 1}

    num = float(org_fig[:-1])
    letter = org_fig[-1].upper()
    
    if letter in keys:
        return num * keys[letter]
    
print(string_tonum('5M'))

5000000.0


In [84]:
popul23[popul23['2005'].astype(str).apply(lambda s: s[-1]).isin(['n', '0'])]

country   1995   1996   1997   1998   1999   2000  2001  2002  2003  \
74   Holy See    NaN    NaN    NaN    NaN    NaN    NaN   NaN   NaN   NaN   
180    Tuvalu  959.0  961.0  963.0  963.0  964.0  964.0  9620  9610  9670   

     2004  2005 2006   2007   2008   2009   2010  
74    NaN   NaN  NaN    NaN    NaN    NaN    NaN  
180  9790  9910  10k  10.1k  10.3k  10.4k  10.6k

--- 
# Part 2

Part 2 requires knowledge of Pandas, EDA, data cleaning, and data visualization.

---

*All libraries used should be added here*

In [4]:
# Imports:

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns



In [79]:
#Data files being used

population = pd.read_csv('../data/population.csv')
life_expect = pd.read_csv('../data/life_expectancy.csv')
health_aid = pd.read_csv('../data/API_SH.XPD.GHED.PP.CD_DS2_en_csv_v2_5872088.csv')
medical_doc = pd.read_csv('../data/medical_doctors_per_1000_people.csv')

## Data Import and Cleaning

### Data Import & Cleaning

Import the datasets that you selected for this project and go through the following steps at a minimum. You are welcome to do further cleaning as you feel necessary:
1. Display the data: print the first 5 rows of each dataframe to your Jupyter notebook.
2. Check for missing values.
3. Check for any obvious issues with the observations (keep in mind reasonable values for each column).
4. Fix any errors you identified in steps 2-3.
5. Display the data types of each feature.
6. Fix any incorrect data types found in step 5.
    - Fix any individual values preventing other columns from being the appropriate type.
    - If your dataset has a column of abbreviated numbers (ex. '3M', '45K', etc.), use the function you wrote in Part 1 (coding challenges, number 3) to convert this to integers! *Hint*: use `.map()` or `.apply()`.
7. Rename Columns.
    - Column names should be all lowercase.
    - Column names should not contain spaces (underscores will suffice--this allows for using the `df.column_name` method to access columns in addition to `df['column_name']`).
    - Column names should be unique and informative.
8. Drop unnecessary rows (if needed).
9. Merge dataframes that can be merged.
10. Perform any additional cleaning that you feel is necessary.
11. Save your cleaned and merged dataframes as csv files.

In [88]:
# Code population cleaning
population.head()
popul23 = population[['country'] + [str(year) for year in range(1995, 2011)]]
missing_pop = popul23[popul23.isnull().any(axis=1)]
print(missing_pop)



Empty DataFrame
Columns: [country, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010]
Index: []


In [89]:
#removing Holy See from dataset as too many dates points missing
population = population.dropna() #dropping Holy See for a lot of missing data
popul23.head()

country   1995   1996   1997   1998   1999   2000   2001  \
0           Afghanistan  16.4M  17.1M  17.8M  18.5M  19.3M  19.5M  19.7M   
1                Angola  13.9M  14.4M  14.9M  15.4M  15.9M  16.4M  16.9M   
2               Albania  3.28M  3.27M  3.25M  3.23M  3.21M  3.18M  3.15M   
3               Andorra  62.9k  64.1k  64.7k  65.2k  65.7k  66.1k  67.8k   
4  United Arab Emirates  2.43M  2.57M  2.75M  2.92M   3.1M  3.28M  3.45M   

    2002   2003   2004   2005   2006   2007   2008   2009   2010  
0    21M  22.6M  23.6M  24.4M  25.4M  25.9M  26.4M  27.4M  28.2M  
1  17.5M  18.1M  18.8M  19.5M  20.2M  20.9M  21.7M  22.5M  23.4M  
2  3.12M  3.09M  3.06M  3.03M     3M  2.98M  2.95M  2.93M  2.91M  
3  70.8k  73.9k  76.9k  79.8k  80.2k  78.2k  76.1k  73.9k  71.5k  
4  3.63M  3.81M  3.99M  4.28M   4.9M  5.87M  6.99M  7.99M  8.48M

In [86]:
popul23.dtypes

country     object
1995       float64
1996       float64
1997       float64
1998       float64
1999       float64
2000       float64
2001        object
2002        object
2003        object
2004        object
2005        object
2006        object
2007        object
2008        object
2009        object
2010        object
dtype: object

In [90]:
#population = population.drop(74) #74 for the row that Holy see was on
#running missing statement again to ensure it has been removed
#popul23.isnull().sum()

#Converting the numbers from abbreviations
for yrs in popul23.columns[1:]:
    popul23[yrs] = popul23[yrs].map(string_tonum)

#popul23.iloc[:, 1:] = popul23.iloc[:, 1:].apply(string_tonum)   

In [91]:
popul23

country        1995        1996        1997        1998  \
0             Afghanistan  16400000.0  17100000.0  17800000.0  18500000.0   
1                  Angola  13900000.0  14400000.0  14900000.0  15400000.0   
2                 Albania   3280000.0   3270000.0   3250000.0   3230000.0   
3                 Andorra     62900.0     64100.0     64700.0     65200.0   
4    United Arab Emirates   2430000.0   2570000.0   2750000.0   2920000.0   
..                    ...         ...         ...         ...         ...   
192                 Samoa    175000.0    177000.0    179000.0    180000.0   
193                 Yemen  16100000.0  16600000.0  17100000.0  17600000.0   
194          South Africa  44000000.0  44700000.0  45300000.0  45900000.0   
195                Zambia   8680000.0   8900000.0   9130000.0   9370000.0   
196              Zimbabwe  11000000.0  11200000.0  11400000.0  11500000.0   

           1999        2000        2001        2002        2003        2004  \
0    19300000.0  19500000.0  19700000.0  21000000.0  22600000.0  23600000.0   
1    15900000.0  16400000.0  16900000.0  17500000.0  18100000.0  18800000.0   
2     3210000.0   3180000.0   3150000.0   3120000.0   3090000.0   3060000.0   
3       65700.0     66100.0     67800.0     70800.0     73900.0     76900.0   
4     3100000.0   3280000.0   3450000.0   3630000.0   3810000.0   3990000.0   
..          ...         ...         ...         ...         ...         ...   
192    182000.0    184000.0    186000.0    187000.0    187000.0    188000.0   
193  18100000.0  18600000.0  19100000.0  19700000.0  20200000.0  20700000.0   
194  46400000.0  46800000.0  47200000.0  47700000.0  48100000.0  48600000.0   
195   9620000.0   9890000.0  10200000.0  10500000.0  10800000.0  11200000.0   
196  11700000.0  11800000.0  11900000.0  12000000.0  12100000.0  12200000.0   

           2005        2006        2007        2008        2009        2010  
0    24400000.0  25400000.0  25900000.0  26400000.0  27400000.0  28200000.0  
1    19500000.0  20200000.0  20900000.0  21700000.0  22500000.0  23400000.0  
2     3030000.0   3000000.0   2980000.0   2950000.0   2930000.0   2910000.0  
3       79800.0     80200.0     78200.0     76100.0     73900.0     71500.0  
4     4280000.0   4900000.0   5870000.0   6990000.0   7990000.0   8480000.0  
..          ...         ...         ...         ...         ...         ...  
192    189000.0    189000.0    190000.0    192000.0    193000.0    195000.0  
193  21300000.0  22000000.0  22600000.0  23300000.0  24000000.0  24700000.0  
194  49000000.0  49500000.0  50000000.0  50600000.0  51200000.0  51800000.0  
195  11600000.0  12000000.0  12400000.0  12900000.0  13300000.0  13800000.0  
196  12200000.0  12300000.0  12500000.0  12600000.0  12700000.0  12800000.0  

[196 rows x 17 columns]

In [ ]:
#Saving the cleaned population file
popul23.to_csv('Population23.csv', index = False)

In [35]:
#Life Expectancy cleaning
life_expect.head()


country  1800  1801  1802  1803  1804  1805  1806  1807  1808  \
0           Afghanistan  28.2  28.2  28.2  28.2  28.2  28.2  28.1  28.1  28.1   
1                Angola  27.0  27.0  27.0  27.0  27.0  27.0  27.0  27.0  27.0   
2               Albania  35.4  35.4  35.4  35.4  35.4  35.4  35.4  35.4  35.4   
3               Andorra   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  United Arab Emirates  30.7  30.7  30.7  30.7  30.7  30.7  30.7  30.7  30.7   

   ...  2091  2092  2093  2094  2095  2096  2097  2098  2099  2100  
0  ...  75.5  75.7  75.8  76.0  76.1  76.2  76.4  76.5  76.6  76.8  
1  ...  78.8  79.0  79.1  79.2  79.3  79.5  79.6  79.7  79.9  80.0  
2  ...  87.4  87.5  87.6  87.7  87.8  87.9  88.0  88.2  88.3  88.4  
3  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
4  ...  82.4  82.5  82.6  82.7  82.8  82.9  83.0  83.1  83.2  83.3  

[5 rows x 302 columns]

In [40]:
#creating new variable with only 1995 to 2010
life_expect23 = life_expect[['country'] + [str(year) for year in range(1995, 2011)]]

In [70]:
#locating missing data
life_expect23[life_expect23.isnull().any(axis=1)]

Empty DataFrame
Columns: [country, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010]
Index: []

In [71]:
#Checking final selected data
life_expect23.head()

country  1995  1996  1997  1998  1999  2000  2001  2002  2003  \
0           Afghanistan  54.3  54.7  54.5  53.3  54.7  54.7  54.8  55.5  56.5   
1                Angola  51.2  51.7  51.6  50.6  51.9  52.8  53.4  54.5  55.1   
2               Albania  74.6  74.5  72.9  74.8  75.1  75.4  76.0  75.9  75.6   
3               Andorra  79.8  80.0  80.2  80.4  80.6  80.8  80.9  81.1  81.2   
4  United Arab Emirates  68.8  68.9  69.0  69.2  69.2  69.1  69.2  69.4  69.3   

   2004  2005  2006  2007  2008  2009  2010  
0  57.1  57.6  58.0  58.5  59.2  59.9  60.5  
1  55.5  56.4  57.0  58.0  58.8  59.5  60.2  
2  75.8  76.2  76.9  77.5  77.6  78.0  78.1  
3  81.3  81.4  81.5  81.7  81.8  81.8  81.8  
4  69.1  69.2  69.5  70.0  70.4  70.6  70.8

In [72]:
#Saving final life expectancy data
life_expect23.to_csv('Life_expectancy.csv', index = False)

In [73]:
#Health aid data cleaning
health_aid.head()

Country Name  1960  1961  1962  1963  1964  1965  1966  \
0                        Aruba   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
1  Africa Eastern and Southern   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
2                  Afghanistan   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
3   Africa Western and Central   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4                       Angola   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

   1967  1968  ...        2013        2014       2015       2016       2017  \
0   NaN   NaN  ...         NaN         NaN        NaN        NaN        NaN   
1   NaN   NaN  ...   87.532252   92.625452  98.537006  97.532557  94.368510   
2   NaN   NaN  ...    8.965855   10.032695  11.157826  12.145635  13.480682   
3   NaN   NaN  ...   29.627769   28.521052  31.641008  29.177849  30.135567   
4   NaN   NaN  ...  129.577242  107.074785  89.779783  84.152683  93.222057   

        2018       2019        2020  2021  2022  
0        NaN        NaN         NaN   NaN   NaN  
1  95.377785  98.107913  102.888434   NaN   NaN  
2  11.656513  10.775585   24.629420   NaN   NaN  
3  30.812392  32.635270   40.102642   NaN   NaN  
4  77.214999  70.747300   77.449551   NaN   NaN  

[5 rows x 64 columns]

In [74]:
#Selecting correct year range to match other data
health_aid20 = health_aid[['Country Name'] + [str(year) for year in range(1995,2011)]]
health_aid20

Country Name  1995  1996  1997  1998  1999        2000  \
0                          Aruba   NaN   NaN   NaN   NaN   NaN         NaN   
1    Africa Eastern and Southern   NaN   NaN   NaN   NaN   NaN   43.326868   
2                    Afghanistan   NaN   NaN   NaN   NaN   NaN         NaN   
3     Africa Western and Central   NaN   NaN   NaN   NaN   NaN   15.069751   
4                         Angola   NaN   NaN   NaN   NaN   NaN   36.498135   
..                           ...   ...   ...   ...   ...   ...         ...   
261                       Kosovo   NaN   NaN   NaN   NaN   NaN         NaN   
262                  Yemen, Rep.   NaN   NaN   NaN   NaN   NaN   65.549442   
263                 South Africa   NaN   NaN   NaN   NaN   NaN  219.173157   
264                       Zambia   NaN   NaN   NaN   NaN   NaN   55.111193   
265                     Zimbabwe   NaN   NaN   NaN   NaN   NaN         NaN   

           2001        2002        2003        2004        2005        2006  \
0           NaN         NaN         NaN         NaN         NaN         NaN   
1     43.616394   43.292173   51.024088   53.532174   56.310836   66.328330   
2           NaN    0.779684    6.277876    5.199076    5.629746    5.538590   
3     19.874533   16.521509   23.263107   29.129152   30.981240   28.343067   
4     69.275662   49.542771   56.252356   71.218851   61.807417   77.386219   
..          ...         ...         ...         ...         ...         ...   
261         NaN         NaN         NaN         NaN         NaN         NaN   
262   70.729858   68.381964   72.663043   61.366065   55.073121   55.844830   
263  204.060650  204.394199  257.033126  267.605490  289.189068  366.687199   
264   56.189848   55.107062   48.018202   50.593274   41.036901   41.521431   
265         NaN         NaN         NaN         NaN         NaN         NaN   

           2007        2008        2009        2010  
0           NaN         NaN         NaN         NaN  
1     72.791742   75.928888   82.645128   82.528924  
2      6.339159    8.420790    8.320093    8.143219  
3     27.206899   26.781339   28.133797   25.676549  
4    104.169230  141.319476  168.316538  110.251719  
..          ...         ...         ...         ...  
261         NaN         NaN         NaN         NaN  
262   55.315939   58.119578   48.159219   45.269081  
263  401.232261  428.057922  471.715236  510.862585  
264   25.281655   15.888849   12.754038   26.385108  
265         NaN         NaN         NaN   49.852982  

[266 rows x 17 columns]

In [75]:
# Identifying all missing data
health_aid20[health_aid20.isnull().any(axis=1)]

Country Name  1995  1996  1997  1998  1999        2000  \
0                          Aruba   NaN   NaN   NaN   NaN   NaN         NaN   
1    Africa Eastern and Southern   NaN   NaN   NaN   NaN   NaN   43.326868   
2                    Afghanistan   NaN   NaN   NaN   NaN   NaN         NaN   
3     Africa Western and Central   NaN   NaN   NaN   NaN   NaN   15.069751   
4                         Angola   NaN   NaN   NaN   NaN   NaN   36.498135   
..                           ...   ...   ...   ...   ...   ...         ...   
261                       Kosovo   NaN   NaN   NaN   NaN   NaN         NaN   
262                  Yemen, Rep.   NaN   NaN   NaN   NaN   NaN   65.549442   
263                 South Africa   NaN   NaN   NaN   NaN   NaN  219.173157   
264                       Zambia   NaN   NaN   NaN   NaN   NaN   55.111193   
265                     Zimbabwe   NaN   NaN   NaN   NaN   NaN         NaN   

           2001        2002        2003        2004        2005        2006  \
0           NaN         NaN         NaN         NaN         NaN         NaN   
1     43.616394   43.292173   51.024088   53.532174   56.310836   66.328330   
2           NaN    0.779684    6.277876    5.199076    5.629746    5.538590   
3     19.874533   16.521509   23.263107   29.129152   30.981240   28.343067   
4     69.275662   49.542771   56.252356   71.218851   61.807417   77.386219   
..          ...         ...         ...         ...         ...         ...   
261         NaN         NaN         NaN         NaN         NaN         NaN   
262   70.729858   68.381964   72.663043   61.366065   55.073121   55.844830   
263  204.060650  204.394199  257.033126  267.605490  289.189068  366.687199   
264   56.189848   55.107062   48.018202   50.593274   41.036901   41.521431   
265         NaN         NaN         NaN         NaN         NaN         NaN   

           2007        2008        2009        2010  
0           NaN         NaN         NaN         NaN  
1     72.791742   75.928888   82.645128   82.528924  
2      6.339159    8.420790    8.320093    8.143219  
3     27.206899   26.781339   28.133797   25.676549  
4    104.169230  141.319476  168.316538  110.251719  
..          ...         ...         ...         ...  
261         NaN         NaN         NaN         NaN  
262   55.315939   58.119578   48.159219   45.269081  
263  401.232261  428.057922  471.715236  510.862585  
264   25.281655   15.888849   12.754038   26.385108  
265         NaN         NaN         NaN   49.852982  

[266 rows x 17 columns]

In [93]:
#dropping any row that has no data at all
health_aid20 = health_aid20.dropna(subset=health_aid20.columns[1:], how='all')

In [94]:
#Filling Na with 0 to make easier to compute
health_aid20 = health_aid20.fillna(0)
health_aid20.head()

Country Name  1995  1996  1997  1998  1999       2000  \
0                        Aruba   0.0   0.0   0.0   0.0   0.0   0.000000   
1  Africa Eastern and Southern   0.0   0.0   0.0   0.0   0.0  43.326868   
2                  Afghanistan   0.0   0.0   0.0   0.0   0.0   0.000000   
3   Africa Western and Central   0.0   0.0   0.0   0.0   0.0  15.069751   
4                       Angola   0.0   0.0   0.0   0.0   0.0  36.498135   

        2001       2002       2003       2004       2005       2006  \
0   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   
1  43.616394  43.292173  51.024088  53.532174  56.310836  66.328330   
2   0.000000   0.779684   6.277876   5.199076   5.629746   5.538590   
3  19.874533  16.521509  23.263107  29.129152  30.981240  28.343067   
4  69.275662  49.542771  56.252356  71.218851  61.807417  77.386219   

         2007        2008        2009        2010  
0    0.000000    0.000000    0.000000    0.000000  
1   72.791742   75.928888   82.645128   82.528924  
2    6.339159    8.420790    8.320093    8.143219  
3   27.206899   26.781339   28.133797   25.676549  
4  104.169230  141.319476  168.316538  110.251719

In [102]:
#Saving file
health_aid20.to_csv('Health aid.csv', index = False)

In [96]:
# Cleaning Medical Doctor data
medical_doc.head()

country   1960  1961  1962  1963  1964   1965  1966  1967  1968  ...  \
0        Aruba    NaN   NaN   NaN   NaN   NaN    NaN   NaN   NaN   NaN  ...   
1  Afghanistan  0.035   NaN   NaN   NaN   NaN  0.063   NaN   NaN   NaN  ...   
2       Angola  0.067   NaN   NaN   NaN   NaN  0.076   NaN   NaN   NaN  ...   
3      Albania  0.276   NaN   NaN   NaN   NaN  0.481   NaN   NaN   NaN  ...   
4      Andorra    NaN   NaN   NaN   NaN   NaN    NaN   NaN   NaN   NaN  ...   

    2010   2011   2012   2013   2014   2015   2016   2017  2018  2019  
0    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   NaN   NaN  
1  0.236  0.252  0.241  0.285  0.298  0.285  0.278    NaN   NaN   NaN  
2    NaN    NaN    NaN    NaN    NaN    NaN    NaN  0.215   NaN   NaN  
3  1.230  1.220  1.270  1.280    NaN    NaN  1.220    NaN  2.16  1.65  
4    NaN    NaN    NaN    NaN    NaN  3.330    NaN    NaN   NaN   NaN  

[5 rows x 61 columns]

In [97]:
#Selecting year range
medical_doc20 = medical_doc[['country'] + [str(year) for year in range(1995,2011)]]
medical_doc20

country   1995   1996    1997  1998   1999   2000  2001  2002  \
0           Aruba  1.120    NaN     NaN   NaN    NaN    NaN   NaN   NaN   
1     Afghanistan    NaN    NaN  0.1100   NaN    NaN    NaN  0.19   NaN   
2          Angola    NaN    NaN  0.0592   NaN    NaN    NaN   NaN   NaN   
3         Albania  1.360  1.380  1.3900  1.39    NaN  1.380  1.31  1.17   
4         Andorra  2.240    NaN  2.4600  2.53    NaN  2.570  2.60  2.90   
..            ...    ...    ...     ...   ...    ...    ...   ...   ...   
202         Samoa    NaN  0.344     NaN  0.33  0.691    NaN   NaN   NaN   
203         Yemen  0.247  0.230  0.2390   NaN    NaN    NaN   NaN   NaN   
204  South Africa    NaN  0.593     NaN   NaN    NaN    NaN   NaN   NaN   
205        Zambia  0.069    NaN     NaN   NaN    NaN    NaN   NaN   NaN   
206      Zimbabwe  0.143    NaN     NaN   NaN    NaN  0.126   NaN   NaN   

      2003    2004    2005    2006    2007    2008    2009    2010  
0      NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN  
1      NaN     NaN     NaN  0.1600  0.1740  0.1740  0.2130  0.2360  
2      NaN  0.0621     NaN     NaN     NaN     NaN  0.1310     NaN  
3    1.190  1.1900     NaN  1.1800  1.2000     NaN  1.2300  1.2300  
4    3.330  3.1300  3.2300  3.0100  3.0100     NaN  3.1500     NaN  
..     ...     ...     ...     ...     ...     ...     ...     ...  
202  0.282     NaN  0.2780     NaN     NaN  0.4640     NaN  0.3440  
203    NaN  0.3450     NaN     NaN     NaN  0.3160  0.3170     NaN  
204    NaN  0.7370     NaN     NaN  0.6990  0.6970  0.7070  0.7210  
205    NaN  0.1300  0.0545  0.0533     NaN  0.0619  0.0606  0.0614  
206    NaN  0.1730  0.1260     NaN  0.0544  0.0596  0.1210  0.1270  

[207 rows x 17 columns]

In [98]:
#Identifying missing data
medical_doc20[medical_doc20.isnull().any(axis=1)]


country   1995   1996    1997  1998   1999   2000  2001  2002  \
0           Aruba  1.120    NaN     NaN   NaN    NaN    NaN   NaN   NaN   
1     Afghanistan    NaN    NaN  0.1100   NaN    NaN    NaN  0.19   NaN   
2          Angola    NaN    NaN  0.0592   NaN    NaN    NaN   NaN   NaN   
3         Albania  1.360  1.380  1.3900  1.39    NaN  1.380  1.31  1.17   
4         Andorra  2.240    NaN  2.4600  2.53    NaN  2.570  2.60  2.90   
..            ...    ...    ...     ...   ...    ...    ...   ...   ...   
202         Samoa    NaN  0.344     NaN  0.33  0.691    NaN   NaN   NaN   
203         Yemen  0.247  0.230  0.2390   NaN    NaN    NaN   NaN   NaN   
204  South Africa    NaN  0.593     NaN   NaN    NaN    NaN   NaN   NaN   
205        Zambia  0.069    NaN     NaN   NaN    NaN    NaN   NaN   NaN   
206      Zimbabwe  0.143    NaN     NaN   NaN    NaN  0.126   NaN   NaN   

      2003    2004    2005    2006    2007    2008    2009    2010  
0      NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN  
1      NaN     NaN     NaN  0.1600  0.1740  0.1740  0.2130  0.2360  
2      NaN  0.0621     NaN     NaN     NaN     NaN  0.1310     NaN  
3    1.190  1.1900     NaN  1.1800  1.2000     NaN  1.2300  1.2300  
4    3.330  3.1300  3.2300  3.0100  3.0100     NaN  3.1500     NaN  
..     ...     ...     ...     ...     ...     ...     ...     ...  
202  0.282     NaN  0.2780     NaN     NaN  0.4640     NaN  0.3440  
203    NaN  0.3450     NaN     NaN     NaN  0.3160  0.3170     NaN  
204    NaN  0.7370     NaN     NaN  0.6990  0.6970  0.7070  0.7210  
205    NaN  0.1300  0.0545  0.0533     NaN  0.0619  0.0606  0.0614  
206    NaN  0.1730  0.1260     NaN  0.0544  0.0596  0.1210  0.1270  

[169 rows x 17 columns]

In [99]:
#dropping any row with no data at all
medical_doc20.dropna(subset=medical_doc20.columns[1:], how='all')

country   1995   1996    1997  1998   1999   2000  2001  2002  \
0           Aruba  1.120    NaN     NaN   NaN    NaN    NaN   NaN   NaN   
1     Afghanistan    NaN    NaN  0.1100   NaN    NaN    NaN  0.19   NaN   
2          Angola    NaN    NaN  0.0592   NaN    NaN    NaN   NaN   NaN   
3         Albania  1.360  1.380  1.3900  1.39    NaN  1.380  1.31  1.17   
4         Andorra  2.240    NaN  2.4600  2.53    NaN  2.570  2.60  2.90   
..            ...    ...    ...     ...   ...    ...    ...   ...   ...   
202         Samoa    NaN  0.344     NaN  0.33  0.691    NaN   NaN   NaN   
203         Yemen  0.247  0.230  0.2390   NaN    NaN    NaN   NaN   NaN   
204  South Africa    NaN  0.593     NaN   NaN    NaN    NaN   NaN   NaN   
205        Zambia  0.069    NaN     NaN   NaN    NaN    NaN   NaN   NaN   
206      Zimbabwe  0.143    NaN     NaN   NaN    NaN  0.126   NaN   NaN   

      2003    2004    2005    2006    2007    2008    2009    2010  
0      NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN  
1      NaN     NaN     NaN  0.1600  0.1740  0.1740  0.2130  0.2360  
2      NaN  0.0621     NaN     NaN     NaN     NaN  0.1310     NaN  
3    1.190  1.1900     NaN  1.1800  1.2000     NaN  1.2300  1.2300  
4    3.330  3.1300  3.2300  3.0100  3.0100     NaN  3.1500     NaN  
..     ...     ...     ...     ...     ...     ...     ...     ...  
202  0.282     NaN  0.2780     NaN     NaN  0.4640     NaN  0.3440  
203    NaN  0.3450     NaN     NaN     NaN  0.3160  0.3170     NaN  
204    NaN  0.7370     NaN     NaN  0.6990  0.6970  0.7070  0.7210  
205    NaN  0.1300  0.0545  0.0533     NaN  0.0619  0.0606  0.0614  
206    NaN  0.1730  0.1260     NaN  0.0544  0.0596  0.1210  0.1270  

[202 rows x 17 columns]

In [100]:
#filling in NaN with 0
medical_doc20 = medical_doc20.fillna(0)
medical_doc20.head()


country  1995  1996    1997  1998  1999  2000  2001  2002  2003  \
0        Aruba  1.12  0.00  0.0000  0.00   0.0  0.00  0.00  0.00  0.00   
1  Afghanistan  0.00  0.00  0.1100  0.00   0.0  0.00  0.19  0.00  0.00   
2       Angola  0.00  0.00  0.0592  0.00   0.0  0.00  0.00  0.00  0.00   
3      Albania  1.36  1.38  1.3900  1.39   0.0  1.38  1.31  1.17  1.19   
4      Andorra  2.24  0.00  2.4600  2.53   0.0  2.57  2.60  2.90  3.33   

     2004  2005  2006   2007   2008   2009   2010  
0  0.0000  0.00  0.00  0.000  0.000  0.000  0.000  
1  0.0000  0.00  0.16  0.174  0.174  0.213  0.236  
2  0.0621  0.00  0.00  0.000  0.000  0.131  0.000  
3  1.1900  0.00  1.18  1.200  0.000  1.230  1.230  
4  3.1300  3.23  3.01  3.010  0.000  3.150  0.000

In [101]:
#Saving file
medical_doc20.to_csv('Medical Doctors.csv', index = False)

In [ ]:



### Data Dictionary

Now that we've fixed our data, and given it appropriate names, let's create a [data dictionary](http://library.ucmerced.edu/node/10249). 

A data dictionary provides a quick overview of features/variables/columns, alongside data types and descriptions. The more descriptive you can be, the more useful this document is.

Example of a Fictional Data Dictionary Entry: 

|Feature|Type|Dataset|Description|
|---|---|---|---|
|**county_pop**|*integer*|2010 census|The population of the county (units in thousands, where 2.5 represents 2500 people).| 
|**per_poverty**|*float*|2010 census|The percent of the county over the age of 18 living below the 200% of official US poverty rate (units percent to two decimal places 98.10 means 98.1%)|

[Here's a quick link to a short guide for formatting markdown in Jupyter notebooks](https://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Working%20With%20Markdown%20Cells.html).

Provided is the skeleton for formatting a markdown table, with columns headers that will help you create a data dictionary to quickly summarize your data, as well as some examples. **This would be a great thing to copy and paste into your custom README for this project.**

*Note*: if you are unsure of what a feature is, check the source of the data! This can be found in the README.

**To-Do:** *Edit the table below to create your own data dictionary for the datasets you chose.*

|Feature|Type|Dataset|Description|
|---|---|---|---|
|column name|int/float/object|This is an example| 


## Exploratory Data Analysis

Complete the following steps to explore your data. You are welcome to do more EDA than the steps outlined here as you feel necessary:
1. Summary Statistics.
2. Use a **dictionary comprehension** to apply the standard deviation function you create in part 1 to each numeric column in one of your dataframes.  **No loops**.
    - Assign the output to variable `sd` as a dictionary where: 
        - Each column name is now a key 
        - That standard deviation of the column is the value
        - *Example Output :* `{'1801': 3.80, '1802': 3.93, ...}`
3. Investigate trends in the data.
    - Using sorting and/or masking (along with the `.head()` method to avoid printing our entire dataframe), consider questions relevant to your problem statement. Some examples are provided below (but feel free to change these questions for your specific problem):
        - Which states have the highest and lowest populations for the past few decades?
        - Which states have the highest and lowest economic growth over the course of the dataset?
        - What years were there new countries created?
        - Do any countries have higher growth rates of population than their economy?
        - Which countries have had the lowest population growth?
    - **You should comment on your findings at each step in a markdown cell below your code block**. Make sure you include at least one example of sorting your dataframe by a column, and one example of using boolean filtering (i.e., masking) to select a subset of the dataframe.

In [ ]:
#Code:

**To-Do:** *Edit this cell with your findings on trends in the data (step 3 above).*

## Visualize the Data

There's not a magic bullet recommendation for the right number of plots to understand a given dataset, but visualizing your data is *always* a good idea. Not only does it allow you to quickly convey your findings (even if you have a non-technical audience), it will often reveal trends in your data that escaped you when you were looking only at numbers. It is important to not only create visualizations, but to **interpret your visualizations** as well.

**Every plot should**:
- Have a title
- Have axis labels
- Have appropriate tick labels
- Text is legible in a plot
- Plots demonstrate meaningful and valid relationships
- Have an interpretation to aid understanding

Here is an example of what your plots should look like following the above guidelines. Note that while the content of this example is unrelated, the principles of visualization hold:

![](https://snag.gy/hCBR1U.jpg)
*Interpretation: The above image shows that as we increase our spending on advertising, our sales numbers also tend to increase. There is a positive correlation between advertising spending and sales.*

---

Here are some prompts to get you started with visualizations. Feel free to add additional visualizations as you see fit:
1. Use Seaborn's heatmap with pandas `.corr()` to visualize correlations between all numeric features.
    - Heatmaps are generally not appropriate for presentations, and should often be excluded from reports as they can be visually overwhelming. **However**, they can be extremely useful in identify relationships of potential interest (as well as identifying potential collinearity before modeling).
    - Please take time to format your output, adding a title. Look through some of the additional arguments and options. (Axis labels aren't really necessary, as long as the title is informative).
2. Visualize distributions using histograms. If you have a lot, consider writing a custom function and use subplots.
    - *OPTIONAL*: Summarize the underlying distributions of your features (in words & statistics)
         - Be thorough in your verbal description of these distributions.
         - Be sure to back up these summaries with statistics.
         - We generally assume that data we sample from a population will be normally distributed. Do we observe this trend? Explain your answers for each distribution and how you think this will affect estimates made from these data.
3. Plot and interpret boxplots. 
    - Boxplots demonstrate central tendency and spread in variables. In a certain sense, these are somewhat redundant with histograms, but you may be better able to identify clear outliers or differences in IQR, etc.
    - Multiple values can be plotted to a single boxplot as long as they are of the same relative scale (meaning they have similar min/max values).
    - Each boxplot should:
        - Only include variables of a similar scale
        - Have clear labels for each variable
        - Have appropriate titles and labels
4. Plot and interpret scatter plots to view relationships between features. Feel free to write a custom function, and subplot if you'd like. Functions save both time and space.
    - Your plots should have:
        - Two clearly labeled axes
        - A proper title
        - Colors and symbols that are clear and unmistakable
5. Additional plots of your choosing.
    - Are there any additional trends or relationships you haven't explored? Was there something interesting you saw that you'd like to dive further into? It's likely that there are a few more plots you might want to generate to support your narrative and recommendations that you are building toward. **As always, make sure you're interpreting your plots as you go**.

In [ ]:
# Code

## Conclusions and Recommendations

Based on your exploration of the data, what are you key takeaways and recommendations? Make sure to answer your question of interest or address your problem statement here.

**To-Do:** *Edit this cell with your conclusions and recommendations.*

**Don't forget to create your README!**

**To-Do:** *If you combine your problem statement, data dictionary, brief summary of your analysis, and conclusions/recommendations, you have an amazing README.md file that quickly aligns your audience to the contents of your project.* Don't forget to cite your data sources!